tools/mirror_releases.sh

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
```

Comment:
Can we remove this file and `versions.bzl` along with it?

protobuf repo is already doing their own releases. I believe we can trust them and don't need to mirror anything.

Reply:
This is not about trust. In order to register prebuilt toolchains we need the information about releases, versions, and their checksum so that we can register a prebuilt protoc like it did before but this time with toolchains instead of hacky select statements. 

forcing users to also use `protobuf` repo just to get the toolchain is not a good design, and has some drawbacks.

- forces users to take protobuf as a dependency even if they don't use any of the lang toolchains there
- protobuf must be careful when making changes to the ruleset as now most bzl files are part of public API contract, in contrary to just making `protoc cli surface` an API contract between rules_proto and protobuf
- compilates the toolchain setup in protobuf, as they now need to declare two types of toolchain, `prebuilt` and `from source`

Here are some examples of versions.bzl pattern from the community and official bazel rules.

https://github.com/bazel-contrib/rules_oci/blob/main/oci/private/versions.bzl
https://github.com/bazel-contrib/rules-template/blob/main/mylang/private/versions.bzl
https://github.com/bazelbuild/rules_python/blob/main/python/versions.bzl
https://github.com/bazelbuild/rules_jvm_external/blob/master/private/versions.bzl
https://sourcegraph.com/github.com/aspect-build/rules_js/-/blob/npm/private/versions.bzl

Reply:
I am happy to make this a follow up though.

Reply:
The design doc https://docs.google.com/document/d/1CE6wJHNfKbUPBr7-mmk_0Yo3a4TaqcTPE0OWNuQkhPs/edit#heading=h.5mcn15i0e1ch
foresees doing registration in protobuf repo.

Main reason here is to remove mainteinance costs and coupling. We don't want to release rules_proto everytime protobuf is release. That's just churn.

I don't see how we're forcing users to depend on protobuf. We are forcing the users to depend on it directly and not transitively. They always need a protoc if they are using `proto_library`, so I'm not following your argument.

Reply:
> Main reason here is to remove mainteinance costs and coupling. We don't want to release rules_proto everytime protobuf is release. That's just churn.

This doesn't make sense. If you are depending on protobuf directly, then you will have to release rules_proto to bump protobuf version anytime there's a new version. There's no other way around this, other than making users patch rules_proto.  

I don't understand the logic behind why cutting a release is a churn. Frequently releases signal a healthy repository and it don't believe it's a churn as long as it's as easy as tagging main branch, and CI takes care of the rest. 

> They always need a protoc if they are using proto_library, so I'm not following your argument.

This is my argument, yes they need always protoc, and it should be provided rules_proto by default as a prebuilt toolchain. If users need a little more complicated setup than prebuilt toolchains, then can remove the `rules_proto_toolchains` call and depend on protobuf repo themselves and call necessary macros to register source toolchains etc..

> Main reason here is to remove maintenance costs and coupling

I don't understand how depending on protobuf reduces the coupling and mainteinance costs. I would argue that depending on protobuf tightly couples two repos as they now don't work without each other.

Reply:
The idea here is that `rules_proto` does not depend on protobuf in the long run, but the other way around: users would import protobuf in their WORKSPACE/bzlmod and that then transitively loads `rules_proto` for `proto_library`. `rules_proto` would only need a release if there are changes to the rules (like `proto_library`), not for the toolchain

Reply:
I’m advocating for removing dependency on protobuf repo and removing versions from rules_proto.

I’m not sure on which of those two points we disagree